File: class06/python/TCC0601-SimuladorAFN.py
#!/usr/bin/env python3
"""TCC0601-SimuladorAFN.py | Aula 06
Simulador de AFN por conjunto de estados ativos.
AFN exemplo: palavras sobre {a,b} que contêm aa ou bb.
"""
from collections import defaultdict

# AFN: δ(q, s) = conjunto de destinos
transicoes = defaultdict(set, {
    ("q0", "a"): {"q0", "q1"},
    ("q0", "b"): {"q0", "q2"},
    ("q1", "a"): {"qf"},
    ("q2", "b"): {"qf"},
    ("qf", "a"): {"qf"},
    ("qf", "b"): {"qf"},
})

q0 = "q0"
finais = {"qf"}


def simular(palavra: str) -> bool:
    """Retorna True se o AFN aceita a palavra (existe caminho vencedor)."""
    ativos = {q0}
    for simbolo in palavra:
        novos = set()
        for q in ativos:
            novos.update(transicoes.get((q, simbolo), set()))
        ativos = novos
        if not ativos:
            return False
    return bool(ativos & finais)


def simular_com_trace(palavra: str):
    """Simula com trace dos conjuntos de estados ativos."""
    ativos = {q0}
    print(f"Início: {{{', '.join(sorted(ativos))}}}")
    for simbolo in palavra:
        novos = set()
        for q in ativos:
            destinos = transicoes.get((q, simbolo), set())
            if destinos:
                print(f"  δ({q}, {simbolo}) = {{{', '.join(sorted(destinos))}}}")
            novos.update(destinos)
        ativos = novos
        print(f"  Após '{simbolo}': {{{', '.join(sorted(ativos))}}}")
        if not ativos:
            print("  → sem estados ativos (REJEITA)")
            return False
    aceita = bool(ativos & finais)
    print(f"Final: {{{', '.join(sorted(ativos))}}} → {'ACEITA' if aceita else 'REJEITADA'}")
    return aceita


if __name__ == "__main__":
    palavras = ["abba", "aba", "aa", "bb", "ab"]
    print(f"{'Palavra':<10} {'Resultado'}")
    print("-" * 25)
    for w in palavras:
        print(f"{w:<10} {'ACEITA' if simular(w) else 'REJEITADA'}")

    print("\n=== Trace detalhado: abba ===")
    simular_com_trace("abba")
    print("\n=== Trace detalhado: aba ===")
    simular_com_trace("aba")

File: class06/java/TCC0601-SimuladorAFN.java
import java.util.*;

/**
 * TCC0601-SimuladorAFN.java | Aula 06
 * Simulador de AFN por conjunto de estados ativos.
 * Mesma lógica e mesma saída do TCC0601-SimuladorAFN.py.
 */
class SimuladorAFN {

    record Par(String estado, String simbolo) {}

    static Map<Par, Set<String>> transicoes = Map.of(
            new Par("q0", "a"), Set.of("q0", "q1"),
            new Par("q0", "b"), Set.of("q0", "q2"),
            new Par("q1", "a"), Set.of("qf"),
            new Par("q2", "b"), Set.of("qf"),
            new Par("qf", "a"), Set.of("qf"),
            new Par("qf", "b"), Set.of("qf")
    );

    static String q0 = "q0";
    static Set<String> finais = Set.of("qf");

    static boolean simular(String palavra) {
        Set<String> ativos = new HashSet<>(Set.of(q0));
        for (char c : palavra.toCharArray()) {
            Set<String> novos = new HashSet<>();
            for (String q : ativos) {
                Set<String> destinos = transicoes.get(new Par(q, String.valueOf(c)));
                if (destinos != null) novos.addAll(destinos);
            }
            ativos = novos;
            if (ativos.isEmpty()) return false;
        }
        return !Collections.disjoint(ativos, finais);
    }

    static void simularComTrace(String palavra) {
        Set<String> ativos = new HashSet<>(Set.of(q0));
        System.out.println("Início: {" + String.join(", ", ativos.stream().sorted().toList()) + "}");
        for (char c : palavra.toCharArray()) {
            String simbolo = String.valueOf(c);
            Set<String> novos = new HashSet<>();
            for (String q : ativos) {
                Set<String> destinos = transicoes.get(new Par(q, simbolo));
                if (destinos != null && !destinos.isEmpty()) {
                    System.out.println("  δ(" + q + ", " + simbolo + ") = {" +
                            String.join(", ", destinos.stream().sorted().toList()) + "}");
                    novos.addAll(destinos);
                }
            }
            ativos = novos;
            System.out.println("  Após '" + simbolo + "': {" +
                    String.join(", ", ativos.stream().sorted().toList()) + "}");
            if (ativos.isEmpty()) {
                System.out.println("  → sem estados ativos (REJEITA)");
                return;
            }
        }
        boolean aceita = !Collections.disjoint(ativos, finais);
        System.out.println("Final: {" + String.join(", ", ativos.stream().sorted().toList()) +
                "} → " + (aceita ? "ACEITA" : "REJEITADA"));
    }

    public static void main(String[] args) {
        List<String> palavras = List.of("abba", "aba", "aa", "bb", "ab");
        System.out.printf("%-10s %s%n", "Palavra", "Resultado");
        System.out.println("-".repeat(25));
        for (String w : palavras) {
            System.out.printf("%-10s %s%n", w, simular(w) ? "ACEITA" : "REJEITADA");
        }

        System.out.println("\n=== Trace detalhado: abba ===");
        simularComTrace("abba");
        System.out.println("\n=== Trace detalhado: aba ===");
        simularComTrace("aba");
    }
}
